plutarch-docs/src/Typeclasses/PLiftable.md

```diff
@@ -0,0 +1,231 @@
+  = DeriveBuiltinPLiftable (a s)
+``
+
+We can see that this `newtype` has two type arguments (besides the `s` to make
+it a Plutarch type): one for the Plutarch type for which we want to derive the
+instance, and one for a Haskell type that is meant to be its `AsHaskell`
+equivalent. To use the derivation helper, you write something like the
+following, substituting `a` and `h` as appropriate:
+
+```haskell
+-- ensure that a is an instance of Plutus type, and h is in the default universe
+
+deriving via (DeriveBuiltinPLiftable a h) instance PLiftable a
+```
+
+### Via `DeriveDataPLiftable`
+
+This helper is designed for types which are represented onchain by way of their
+`Data` encoding, rather than being part of the Plutus universe directly. An
+example of such a type is `PScriptContext`: its equivalent in Haskell is
+`ScriptContext` from `plutus-ledger-api`, which is essentially a 'skin' over
+`Data`.
+
+This type is defined as follows: its implementation is not important.
+
+```haskell
+newtype DeriveDataPLiftable (a :: S -> Type) (h :: Type) (s :: S)
+  = DeriveDataPLiftable (a s)
+```
+
+Similarly to `DeriveBuiltinPLiftable`, we have two relevant type arguments: 
+one for the Plutarch type for which we want to derive an instance, and 
+another for its Haskell-level equivalent. Much as with `DeriveBuiltinPLiftable`,
+we use `DeriveDataPLiftable` as so, substituting `a` and `h` appropriately:
+
+```haskell
+deriving via (DeriveDataPLiftable a h) instance PLiftable a
+```
+
+There are more stringent requirements of `a` and `h` for this to work. Aside
+from `a` being an instance of `PlutusType`, we also must ensure that:
+
+* `PInner a` is `PData` (namely, construction and matching is actually carried
+  out in a computation involving `Data`)
+* `h` is an instance of both `ToData` and `FromData` (namely, it has a `Data`
+  encoding that we can decode from and encode into)
+
+### Manual derivation
+
+In the (unlikely) case that your type fits neither of the above, you will have
+to write the instance manually. Given how unusual this situation is, we can't
+really give any general guidance as to how this should be done. Instead, we
+recommend examining the definitions of `PLiftable`, as well as the two
+derivation helpers described above, in the source code. Alternatively, reach out
+to us: we might be able to advise better.
+
+### Ensuring your instance is correct
+
+Any instance of `PLiftable` must obey the following laws:
+
+```
+1. fromPlutarch . toPlutarch = Right
+2. fmap toPlutarch . fromPlutarch = Right
+```
+
+If you use either of `DeriveBuiltinPLiftable` or `DeriveDataPLiftable` with
+`via` derivation, these laws will automatically be satisfied. If you write an
+instance manually, you will have to ensure this yourself. We provide a helper
+for testing that these laws hold in `plutarch-testlib`, in the
+`Plutarch.Test.Laws` module, called `checkPLiftableLaws`, which uses QuickCheck
```

Comment:
Let's show how to define test case for the law. Feel like this needs more emphasis